File: find_minimum_element.py
"""
For finding the minimum element we basically just need to find the pivot element of the array
Whichever side of mid is sorted, minimum will never be on that side because the aray is rotated sorted array
1. Find the sorted side
2. If rotated side is left then ignore it and move to right of mid
3. If rotated side is right then ignore it and move to left of mid
4. If mid is lower than mid-1 and mid+1 then minimum is mid, so return mid
"""

# Time complexity - O(log N)
# Space complexity - O(1)

class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """        
        if len(nums) == 0 or nums is None:
            return -1
        
        low = 0
        high = len(nums) - 1
        
        while low <= high:
            mid = low + (high - low)/2
            
            if nums[low] < nums[high]:
                return nums[low]
            
            if (mid == 0 or nums[mid] < nums[mid-1]) and (mid == len(nums)-1 or nums[mid] < nums[mid+1]):
                return nums[mid]
            
            # left side is sorted
            elif nums[low] <= nums[mid]:
                low = mid + 1
                
            # right side is sorted
            else:
                high = mid - 1 
                
        return -1
                                
            

File: FindMinimumElement.java
/*
 *  Find Minimum in Rotated Sorted Array
 */

class FindMinimumElement {
    public int findMin(int[] nums) {
        if(nums == null || nums.length == 0){
            return -1;
        }

        int low = 0;
        int high = nums.length - 1;

        while(low <= high){
            int mid = low + (high - low)/2; //condition to check integer overflow
            
            //this condition means that array is sorted
            if(nums[low] <= nums[mid] && nums[high] > nums[mid]){
                return nums[low];
            }

            if ((mid == 0 || nums[mid - 1] > nums[mid]) && (mid == nums.length - 1 || nums[mid + 1] > nums[mid])) {
                return nums[mid];
            }

            //check if left side is sorted, chances are minmum number will be found in unsorted array
            if(nums[low] <= nums[mid]){
                low = mid + 1;
            }else{
                high = mid - 1;
            }
        }
        return nums[low];
    }
    public static void main(String[] args) {
        FindMinimumElement solution = new FindMinimumElement();

        // Test case 1: Example test case
        int[] nums1 = {3,4,5,1,2};
        int result1 = solution.findMin(nums1);
        System.out.println("Test Case 1:");
        System.out.println("Expected Output: 1");
        System.out.println("Actual Output: " + result1);

        // Test case 2
        int[] nums2 = {1, 2, 3, 4, 5};
        int result2= solution.findMin(nums2);
        System.out.println("\nTest Case 2:");
        System.out.println("Expected Output: 1");
        System.out.println("Actual Output: " + result2);

        // Test case 3
        int[] nums3 = {1};
        int result3 = solution.findMin(nums3);
        System.out.println("\nTest Case 3:");
        System.out.println("Expected Output: 1");
        System.out.println("Actual Output: " + result3);

}
}